Allow the generic animation classes to be used on all platforms

RobinD42 opened the pull request:
* Add `--disable-nativeanimation` configure option, and appropriate conditionals to support using the generic wxAnimation and wxAnimationCtrl classes on all platforms
* Update the interface headers to include the generic-only methods
* Add the interface headers for the wxAnimationDecoder classes

Commit:
Add --disable-nativeanimation configure option, and wxUSE_NATIVE_ANIMATIONCTRL

Commit:
Move generic animate files to GUI_CMN

Commit:
Compile native or generic animation classes based on wxUSE flags

Commit:
Run bakefile

RobinD42 committed:
DoCanRead is protected in the base class, make it the same in derived classes.

RobinD42 committed:
Update the interface files for the animation classes

RobinD42 committed:
Fix interface filename, remove duplicate enum

RobinD42 committed:
Split decoder interface to match the actual header files

vadz committed:
Revert "Run bakefile"

This reverts commit fea3d52505bdd7103052b42c5e9c92ee3b67ebeb.

vadz commented:
I'm really not sure what happened to the makefiles here... I've reverted fea3d52505 to be able to view the real differences in this PR, I'll rerun bakefile myself before merging.

RobinD42 commented:
Okay, thanks.

vadz commented on include/wx/setup_inc.h:
```diff
@@ -919,6 +919,15 @@
 #define wxUSE_NATIVE_DATAVIEWCTRL 1
 
+// Use generic version of wxAnimationCtrl and supporting classes, even if a 
+// native one is available?
```
This seems reversed. I.e. the description corresponds to the value of 0 of this option or something like that.

vadz commented on setup.h_vms:
```diff
@@ -465,6 +465,8 @@ typedef pid_t GPid;
 
 #define wxUSE_NATIVE_DATAVIEWCTRL 0
 
+#define wxUSE_NATIVE_ANIMATIONCTRL 0
```
It might be better to set this to 1 by default?

vadz commented on src/gtk/animate.cpp:
```diff
@@ -12,9 +12,10 @@
 #include "wx/wxprec.h"
 
 #if wxUSE_ANIMATIONCTRL && !defined(__WXUNIVERSAL__)
```
```suggestion
#if wxUSE_ANIMATIONCTRL
```
I don't think we need this check here any longer, as it's done in the header.

vadz reviewed (commented):
No real comments, but I wonder what limitations of the native version pushed you to do this? Maybe they should be documented somewhere?

Otherwise the only thing I'd like to fix is the `wxUSE_NATIVE_ANIMATIONCTRL` option description.

TIA!

RobinD42 commented:
> I wonder what limitations of the native version pushed you to do this?

The main thing was the inability to add new decoders for new file types. Secondarily was that I couldn't wrap the generic methods for wxPython that were not present or used in the native GTK version of the classes.

RobinD42 committed:
Remove check for __WXUNIVERSAL__

Co-Authored-By: VZ <vadim@wxwidgets.org>

RobinD42 committed:
Update description of wxUSE_NATIVE_ANIMATIONCTRL

RobinD42 committed:
Add text to wxAnimation docs about how to use the generic version on wxGTK

RobinD42 commented:
Okay, I'm done with my updates.

MaartenBent commented:
Also `build/cmake/setup.h.in` is not updated. It should automatically update when running `update-setup-h`.

vadz committed:
Remove extra accidentally added blank line in auto-generated file

Just commit the result after running build/update-setup-h.

vadz committed:
Update CMake version of setup.h header too

This is the result of running build/update-setup-h too.

vadz committed:
Rerun bakefile after wxAnimationCtrl-related changes

vadz commented:
I've updated the CMake file (thanks Maarten!), I think Robin might not have known about `build/update-setup-h` helper existence as another file was apparently updated manually too and rerun bakefile to give the CI builds a chance now.

If they do pass, I'm going to merge this because I can't see any real harm in it, but thinking more about this, I realize that a much better solution would be to implement _runtime_ fallback from the native version to the generic one, similarly to how it's done for e.g. `wxHyperlinkCtrl`. The implementation of `UseNative()` there is based on the GTK version, but we could also do it depending on the requested format and/or if some flag is specified. The main advantage of this would be to allow using the standard build of wx in wxPython because I don't really understand how is this going to work for Linux packages otherwise -- surely we want wxPython packages to use the system wx packages, don't we?

MaartenBent commented:
Similar issue for `build/cmake/files.cmake`. `build/files` should be updated and used to generate the bkl and cmake file.

vadz committed:
Move generic files to common sources/headers in the right place

This corresponds to the changes already manually done in
build/bakefiles/files.bkl but is now done in the right file and so also
for CMake-based build system.

vadz committed:
Remove the now unused UNIV_PLATFORM_{SRC,HDR} variables

They used to contain animation-related files, but are empty now and can
be simply removed.

RobinD42 commented:
>  I think Robin might not have known about build/update-setup-h

That's correct. It's been a very long time since I had to make changes to any setup.h or bakefiles myself, and if I did know about it then I've since forgotten. Maybe this time it will stay in my memory...

Yes, a runtime fallback would be good. Another alternative would be to expose the generic version of the classes as wxGenericAnimation and wxGenericAnimationCtrl on all platforms, and then use typedefs or #defines for the platforms without a native implementation to let them use the non-generic names. IIRC this is how it's done for wxCalendarCtrl.  WIth this approach the programmers can use wxAnimation[Ctrl] if they don't care about the extra functionality, or wxGenericAnimation[Ctrl] if they do.

I do have a few tricks for dealing with differences between classes in different builds of wxPython. But in this case I haven't quite worked out how to do it yet, since there's a bunch of virtuals and associated functionality missing between the two implementations.

vadz commented:
> Yes, a runtime fallback would be good. Another alternative would be to expose the generic version of the classes as wxGenericAnimation and wxGenericAnimationCtrl on all platforms, and then use typedefs or #defines for the platforms without a native implementation to let them use the non-generic names. 

This is a prerequisite for what I proposed, as the native `wxAnimationCtrl` would need to inherit from `wxAnimationCtrlGeneric` in order to be able to fall back on it.

But, just to be clear, I'd be happy with anything which allows to avoid having to build a non-standard wx version for wxPython which is something I'd really like to avoid.

Anyhow, should I merge this or will you put something else in place instead/on top of it in this branch?

RobinD42 commented:
I'll spend some more time on it and see what I can come up with.

RobinD42 committed:
Remove --disable-nativeanimation and wxUSE_NATIVE_ANIMATIONCTRL

RobinD42 committed:
Run autoconf

RobinD42 committed:
* Rename the generic version as wxGenericAnimationCtrl
* Add a simple shim class for the generic wxAnimationCtrl
* Use the generic wxAnimation as the base class (this will probably change...)

RobinD42 committed:
Name the base animation class wxGenericAnimation with a wxAnimation shim class

RobinD42 committed:
Name the base animation class wxGenericAnimation with a wxAnimation shim class

RobinD42 committed:
* Switch back to using an wxAnimationBase class
* Change the [GS]etAnimation methods to be non-virtual so they can use generic/native types of animation obj

Commit:
native wxAnimationCtrl derives from the generic class

RobinD42 committed:
Add a Create method to the shim class

RobinD42 committed:
Documentation updates for Animation classes

RobinD42 committed:
IsOk is pure virtual

RobinD42 commented:
Okay, I'm not completely satisfied with the result but I've shifted things around so both the generic and native classes are available all the time, and it works well from wxPython too.

I had hoped to be able to find a way to have just one `wxAnimation` class for both types of ctrl, but it wasn't clear to me how to deal with the gtk-specific things in that case. So now there is both `wxGenericAnimation` and `wxAnimation` which isn't too bad, but that also means that there is both the `wxNullGenericAnimaiton` and `wxNullAnimation` constants, which is kind of ugly. It also meant that the `GetAnimation` and `SetAnimation` in the ctrl classes needed to be made non-virtual and to have slightly different implementations.

If you can see any better ways to deal with those issues please go ahead and do so.

vadz commented on include/wx/animate.h:
```diff
@@ -21,11 +21,19 @@
+// ----------------------------------------------------------------------------
+
+WX_DECLARE_LIST_WITH_DECL(wxAnimationDecoder, wxAnimationDecoderList, class WXDLLIMPEXP_ADV);
+
```
It doesn't matter much, but I don't see why did this have to be moved here. AFAICS it's only used in the generic header and could remain there. Couldn't it?

RobinD42 replied:
I moved things around several times. That line was probably just left behind when things moved back.

vadz commented on include/wx/generic/animate.h:
```diff
@@ -14,19 +14,17 @@
         { LoadFile(name, type); }
 
-    virtual bool IsOk() const wxOVERRIDE
+    virtual bool IsOk() const 
```
Why did `override` disappear? AFAICS it should still be there (and normally its absence should result in clang warnings).

RobinD42 replied:
Not sure, it may have been part of the refacter that was then missed when things mostly changed back.

vadz commented on include/wx/animate.h:
```diff
@@ -49,6 +57,13 @@ class WXDLLIMPEXP_CORE wxAnimationBase : public wxObject
+    virtual wxAnimationDisposal GetDisposalMethod(unsigned int frame) const = 0;
+    virtual wxColour GetTransparentColour(unsigned int frame) const = 0;
+    virtual wxColour GetBackgroundColour() const = 0;
+
```
Sorry, I don't understand why did we have to move all these methods into the base class. It looks like they're only used in the generic code and generic code should always work with generic animation objects, so it really should be enough to have these methods there only.

Am I missing something?

RobinD42 replied:
I was thinking that it might make sense to allow custom classes derived from `wxAnimationBase` in some (likely rare) cases. That doesn't make as much sense with the `Impl` model, and it would probably be better with custom decoder classes instead in either case.

vadz commented on src/generic/animateg.cpp:
```diff
@@ -14,8 +14,7 @@
 
-#if wxUSE_ANIMATIONCTRL
-
+#if wxUSE_ANIMATIONCTRL 
```
```suggestion
#if wxUSE_ANIMATIONCTRL 

```
just to avoid a spurious chunk in the diff.

vadz commented on src/gtk/animate.cpp:
```diff
@@ -11,10 +11,10 @@
+#if wxUSE_ANIMATIONCTRL
 #include "wx/animate.h"
 
+
```
```suggestion
```

vadz commented on src/gtk/animate.cpp:
```diff
@@ -11,10 +11,10 @@
 
-#if wxUSE_ANIMATIONCTRL && !defined(__WXUNIVERSAL__)
-
+#if wxUSE_ANIMATIONCTRL
```
Not as if it were really important, but I'm not sure why wxUniv check got removed from here?

RobinD42 replied:
You had suggested the same change in `animation.h` so I made it here too.

vadz reviewed (commented):
Thanks for the updates!

I agree that this is much better now, but working by value with a polymorphic class (`wxAnimation`) is never going to be pretty. It was fine before because it wasn't effectively polymorphic from the user point of view, as they never dealt with other members of the same hierarchy, but not it's unavoidable.

There are just 2 options here: either use (smart) pointers instead or make the class itself a kind of smart pointer. I think the second option would be much preferable here, I.e. instead of inheriting `wxAnimation` and `wxGenericAnimation` from `wxAnimationBase`, we should remove the latter to `wxAnimation` itself and add `wxAnimationImpl* m_impl` to it. Then rename the other 2 classes to `wxAnimationGTKImpl` and `wxAnimationGenericImpl` respectively.

This would allow us to keep only a single `wxNullAnimation` (equivalent of null pointer) and continue having `wxAnimation GetAnimation()` in all classes, even the generic one in GTK build, which is also important.

Do you think you could make a last effort and re-refactor things like this? Of course, please let me know if you have any questions (or if I'm missing anything). And TIA!

RobinD42 commented on include/wx/animate.h:
```diff
@@ -114,10 +126,57 @@ class WXDLLIMPEXP_CORE wxAnimationCtrlBase : public wxControl
+    #include "wx/generic/animate.h"
     #include "wx/gtk/animate.h"
-#else
+#else  
```
Looks like I had the editor's auto-trim turned off on my linux machine. I'll clean those up.

RobinD42 commented on include/wx/animate.h:
```diff
@@ -114,10 +126,57 @@ class WXDLLIMPEXP_CORE wxAnimationCtrlBase : public wxControl
 
-#if defined(__WXGTK20__) && !defined(__WXUNIVERSAL__)
+#if defined(__WXGTK20__)
+    #include "wx/generic/animate.h"
```
Yep. It was that way at one point, so this is probably another casualty of moving things around a few times.

RobinD42 committed:
commit suggested change

Co-Authored-By: VZ <vadim@wxwidgets.org>

RobinD42 commented:
> or make the class itself a kind of smart pointer

I should have thought about that. I just explained the similar pattern in the `wxDC` family to somebody a few weeks ago.

What's your suggestion for how/where the `Impl` instance should be created? Should the app developer create it and pass it to `wxAnimation`?  Should there be a generic/native flag? A separate factory function of some sort?

RobinD42 commented:
Oops, wrong button!

RobinD42 committed:
commit suggested change

Co-Authored-By: VZ <vadim@wxwidgets.org>

RobinD42 committed:
Trim trailing whitespace

vadz commented:
> What's your suggestion for how/where the `Impl` instance should be created? Should the app developer create it and pass it to `wxAnimation`?

I didn't think about allowing custom animation implementations at all, but if you want to allow this, we would indeed need a way to do something like this. However I'm not sure if/why do we really want/need this? Perhaps we could postpone doing this, as it looks like this could be always added later while preserving compatibility?

> Should there be a generic/native flag and the `wxAnimation` creates the `Impl`? A separate factory function of some sort?

AFAICS animations are actually only really created by `Load()` method, so it's tempting to base the decision of which "Impl" to create based on `wxAnimationType` passed to it, but I guess this will be insufficient, as you may want to use generic animation even for GIFs, which are supported by the native version, right? Of course, we could add `wxANIMATION_TYPE_GIF_GENERIC` for this, but it looks a bit ugly. OTOH it's probably the simplest solution.

A factory function would indeed work and a natural place for it would be `wxAnimationCtrl` itself. I.e. we could add `CreateAnimation()` (documented as "creating an animation object compatible with this control") that would set things up appropriately. Or maybe we could turn this around and add `wxAnimationCtrl *controlToBeUsedWith = NULL` to `wxAnimation` ctors (except the copy one) that would call a private (in the sense not part of the public API) `wxAnimationCtrl::CreateAnimationImpl()` internally to create the right impl object?

I think I finally like the latter solution the most, but I could live with either of them -- or with something better if you see any way to improve this, of course.

RobinD42 committed:
* Switch wxAnimation to be a facade over a wxAnimationImpl class.
* Implement wxAnimationGenericImpl
* Update wxGenericAnimationCtrl as needed

RobinD42 committed:
Verify the animation impl type matches the animation ctrl type

RobinD42 committed:
Ensure the animation is ok before getting its impl

RobinD42 committed:
Remove the generic-specific methods from the wxAnimation API

RobinD42 committed:
Update docs

RobinD42 committed:
Remove the wxOVERRIDEs too

Commit:
Update GTK animation classes for the new pattern

RobinD42 committed:
Add missing wxOVERRIDE, and some other cleanup

RobinD42 committed:
More documentation updates

RobinD42 commented:
@vadz this is ready for review again.

vadz commented on src/gtk/animate.cpp:
```diff
@@ -284,8 +253,19 @@ void wxAnimationCtrl::SetAnimation(const wxAnimation &anim)
     ResetAnim();
     ResetIter();
 
+    m_animation = anim;
```
Why do we use `m_animation` here? This is the base (generic) class member, so I think we shouldn't touch it at all when we're not using the generic implementation.

vadz replied:
It still feels a bit dirty to reuse the base class object like this here, but I guess it's ok then... To preserve what the current code does we'd need to return `wxAnimation(new wxAnimationImplGTK(m_anim))`, I guess?

RobinD42 replied:
Yes, or something like that.

vadz commented on src/gtk/animate.cpp:
```diff
@@ -466,4 +446,36 @@ void wxAnimationCtrl::OnTimer(wxTimerEvent& WXUNUSED(ev))
+        case wxANIMATION_IMPL_TYPE_NATIVE:
+            return new wxAnimationGTKImpl();
+
+        default:
+            return NULL;
```
Minor but it would be better to remove `default:` and move `return NULL` below the switch.

The rationale is that the compiler would then give us a warning here if we ever add a new element to the enum over which we switch, while it won't if there is a default case.

vadz reviewed (commented):
Thanks, I don't see any real problems remaining, but I have 2 questions/remarks it might be worth to address:

1. Do we really want to make the `Impl` class public? I'd prefer to hide them in `include/wx/private/animate.h` and not expose them at all as I don't think it's useful, there doesn't seem to be any way to create user-defined defined implementation classes in `wxAnimation` anyhow.
1. I'm not sure if `wxAnimation` objects were supposed to be copyable, but it seems like there was some code in wxGTK version for this, which got removed. Maybe this is fine, and I don't really see why would anyone want to copy animations anyhow, but I just wanted to check if this was intentional.

Thanks again!

RobinD42 committed:
Apply suggestions from code review

Co-Authored-By: VZ <vadim@wxwidgets.org>

RobinD42 committed:
Fix class names in the header

RobinD42 committed:
Revert change that ended up just moving some lines

RobinD42 committed:
Use wxDECLARE_NO_COPY_CLASS in the Impl classes

RobinD42 committed:
Use better names for the Impl accessor methods

RobinD42 committed:
Remove default case from switch

RobinD42 committed:
No need for wxRTTI in wxAnimationImpl

RobinD42 committed:
Minor documentation tweaks

RobinD42 commented:
> Do we really want to make the Impl class public?

I was about to change this when I realized the if we want to allow custom `wxAnimationDecoder`s (we do, it's one of the reasons I started all this) then currently we need to keep `wxAnimationGenericImpl` public because that is where the `AddHandler` and related static methods are located. Would it make more sense for those to be located in `wxAnimationDecoder`? Or maybe someplace else?

> I'm not sure if wxAnimation objects were supposed to be copyable

`wxAnimation` is copyable, although I see that there isn't an assignment operator. I'll add that.  The `Impl` objects do not need to be copyable, and they are ref counted so I think they should do the right thing when the `wxAnimation` is copied.

RobinD42 committed:
Add assignment operator to wxAnimation

Commit:
wxAnimationGTKImpl::GetFrame got lost somwhere

vadz commented:
> > Do we really want to make the Impl class public?
> 
> I was about to change this when I realized that if we want to allow custom `wxAnimationDecoder`s (we do, it's one of the reasons I started all this) then currently we need to keep `wxAnimationGenericImpl` public because that is where the `AddHandler` and related static methods are located. Would it make more sense for those to be located in `wxAnimationDecoder`? Or maybe someplace else?

I think either `wxAnimationDecoder` or `wxAnimation` itself would be fine. It's not ideal that custom decoders wouldn't be usable by the native GTK implementation, of course, but I don't really see how to avoid this.

And in principle it's possible to transcode the data on the fly in the native version...

RobinD42 committed:
Move wxAnimationDecoderList-related methods to wxAnimation

RobinD42 committed:
Move the animation Impl classes to private headers

Commit:
gtk animation ctrl needs to see the generic impl too

RobinD42 committed:
Documentation updates

vadz commented:
Thanks for all the updates! This looks good to me, should I merge it now?

RobinD42 commented:
Yes, it's ready. Thanks.

vadz committed:
Restore blank line present in master to minimize diff

No real changes.

vadz committed:
Use WXDLLIMPEXP_CORE instead of WXDLLIMPEXP_ADV in new code

They're the same thing, and we still keep WXDLLIMPEXP_ADV in the
existing code, but as this line was just moved to a new place, update it
to use the more appropriate DLL export declaration too.

No real changes.

vadz committed:
Remove some recently added trailing spaces

No changes.

vadz committed:
Define wxHAS_NATIVE_ANIMATIONCTRL

This allows to easily test for the native control existence and also
whether wxGenericAnimationCtrl is really different from wxAnimationCtrl.

It also allows to avoid explicit check for wxGTK in common code.

vadz committed:
Add possibility to use generic version to the animation sample

Add menu item to switch to the generic version when using the sample on
a platform where a native version is available (i.e. wxGTK) in order to
allow testing it there easily.

vadz commented:
Thanks, I've added a couple of tiny and insignificant fixes and another somewhat more important commit allowing to use the generic version in the sample, please let me know if you have any comments about it (I must admit that I'm not terribly happy about creating generic `wxAnimation` object, I think we might want to add some better way to do it).

I also have a new question: why do we derive GTK native control from `wxGenericAnimationCtrl`? It looks like we don't really need to do it, do we?

vadz committed:
Remove wxAnimationCtrlBase from wxRTTI

This is consistent with the other wxFooBase classes which are not part
of it, as they're considered implementation details and not parts of the
API.

vadz commented:
Oops, sorry, and another fix, to wxRTTI. Again, please let me know if you have any objections. TIA!

RobinD42 commented:
> I also have a new question: why do we derive GTK native control from wxGenericAnimationCtrl? It looks like we don't really need to do it, do we?

No, probably not. I was just following the example of wxHyperlinkCtrl

RobinD42 commented:
If. I tapped the wrong button again

vadz commented:
> > I also have a new question: why do we derive GTK native control from wxGenericAnimationCtrl? It looks like we don't really need to do it, do we?
> 
> No, probably not. I was just following the example of wxHyperlinkCtrl

OK, I'll change this too then. `wxHyperlinkCtrl` needs to do it because it can fall back on the generic version during run-time if GTK versions it too low. Here we don't have this problem.

RobinD42 commented:
Ok, thanks.

vadz commented:
Looking at the GTK code now I'm afraid we ended up with something rather strange here, as I see no reason creating `wxAnimationGenericImpl` could be useful/needed in its `CreateAnimationImpl()`... It looks like this is a compromise/mix between the various solutions we discussed, but it doesn't really make much sense to me as it now. I'll try to do something about this too, but it might take a bit longer.

vadz committed:
Don't derive wxAnimationImpl from wxObject

This is just unnecessary and having wxAnimation::m_refData->m_refData is
confusing, both in wxGTK version where it's not used and in the generic
one where it is, but can be replaced by more type-safe m_decoder.

vadz committed:
Make wxAnimationImpl private and get rid of wxAnimationImplType

Simplify and streamline animation classes relationship: wxAnimation is
the only public class representing an animation and it can be created by
both the native wxAnimationCtrl and wxGenericAnimationCtrl using the new
public CreateAnimation() method.

Replace wxAnimationImplType enum with more flexible type info based
check.

vadz committed:
Move wxGenericAnimationCtrl::m_animation to the base class

API of wxAnimationCtrlBase practically forces the derived classes to
have such method, so just add it and make GetAnimation() a simple
non-virtual accessor.

vadz committed:
Stop deriving native wxGTK wxAnimationCtrl from generic one

This is unnecessary and confusing.

vadz committed:
Document wxGenericAnimationCtrl as deriving from wxAnimationCtrl

Even though this is not really the case at the code level, this makes
more sense, as wxGenericAnimationCtrl adds methods to the interface
implemented by wxAnimationCtrl.

This also allows to avoid having notes indicating that some methods
don't work for the native version of the control -- by simply not having
them there.

vadz commented:
Sorry, I ended up doing some relatively major changes, but I believe the end result is much better.

The gist of them is that I've made `wxAnimationImpl` really private and removed `wxAnimationImplType` completely. Instead, `wxAnimation` objects compatible with the control using them can be created using its `CreateAnimation()` method and this simplifies the sample code, which is nicer now.

I've also stopped deriving GTK native control from the generic one and updated the documentation to not say this.

Please let me know if I've simplified too much.

vadz committed:
Restore lost check for __WXUNIVERSAL__ in wx/animate.h

Don't use native version when building wxUniv.

vadz committed:
Reduce insignificant documentation changes with master

Restore the original amount of whitespace and remove WXDLLIMPEXP_CORE
which shouldn't have appeared in the documentation header in the first
place.

RobinD42 commented:
The fact that wxGenericAnimationCtrl has different parents in different builds complicates things a little for wxPython, but I can work around that. (And already have in a few other cases.) 

But there is one thing is not clear. In a wxGTK build how do I create a `wxAnimation` that I can pass to the `wxGenericAnimationCtrl` constructor? For example, I have code like this in a wxPython demo:

```
        ani = wx.Animation(gif_name)
        ctrl = wx.AnimationCtrl(self, -1, ani)
        ctrl.Play()
```

If I change that to a `wx.GenericAnimationCtrl` then I trigger the assert at animateg.cpp line 273.

vadz commented:
> The fact that `wxGenericAnimationCtrl` has different parents in different builds

But it doesn't, actually? It always inherits from `wxAnimationCtrlBase`. `wxAnimationCtrl` itself can or not inherit from `wxGenericAnimationCtrl`, but this shouldn't be a problem?

> But there is one thing that is not clear. In a wxGTK build how do I create a `wxAnimation` that I can pass to the `wxGenericAnimationCtrl` constructor? 

You create it by using this control itself, see animate sample. If this is too restrictive because you need an instance of a control in order to do it, we can easily add a static class method doing this too.

> For example, I have code like this in a wxPython demo:
> 
> ```
>         ani = wx.Animation(gif_name)
>         ctrl = wx.AnimationCtrl(self, -1, ani)
>         ctrl.Play()
> ```
> 
> If I change that to a `wx.GenericAnimationCtrl` then I trigger the assert at animateg.cpp line 273.

This should work right now:
```python
ctrl = wx.GenericAnimationCtrl(self, -1)
ani = ctrl.CreateAnimation()
ani.Load(gif_name)
ctrl.SetAnimation(ani)
ctrl.Play()
```

With my suggestion above we could do:
```python
ani = wx.GenericAnimationCtrl.CreateAnimation()
ani.Load(gif_name)
ctrl = wx.GenericAnimationCtrl(self, -1, ani)
ctrl.Play()
```

It's still a bit awkward, but in this simple case it could, of course, be simplified to just
```python
ctrl = wx.GenericAnimationCtrl(self, -1)
ctrl.LoadFile(gif_name)
ctrl.Play()
```
without using animation object explicitly at all.

RobinD42 commented:
> > The fact that wxGenericAnimationCtrl has different parents in different builds
> 
> But it doesn't, actually? It always inherits from wxAnimationCtrlBase. wxAnimationCtrl itself can or not inherit from wxGenericAnimationCtrl, but this shouldn't be a problem?

Keep in mind that the wrapper generator for wxPython uses the interface headers to define the classes and relationships between classes (the hierarchy and other info, including how to cast from one type to another.) The interface headers give us this:

```
class wxAnimationCtrl : public wxControl {...}
class wxGenericAnimationCtrl : public wxAnimationCtrl {...}
```

which is different from the reality, but apparently is close enough to reality in a generic-only build to work fine with wxPython's generated code. But in a GTK build this is the reality:

```
class wxAnimationCtrl: public wxAnimationCtrlBase {...}
class wxGenericAnimationCtrl: public wxAnimationCtrlBase {...}
```

I end up with this error in the build:

```
../../../../sip/cpp/sip_advwxGenericAnimationCtrl.cpp: In function ‘void* cast_wxGenericAnimationCtrl(void*, const sipTypeDef*)’:
../../../../sip/cpp/sip_advwxGenericAnimationCtrl.cpp:2715:56: error: invalid static_cast from type ‘wxGenericAnimationCtrl*’ to type ‘wxAnimationCtrl*’
         return static_cast< ::wxAnimationCtrl *>(sipCpp);
                                                        ^
```

Hmm... Now that I look closer at it close I see that there's a potential hidden problem on generic-only builds too since `wxAnimationCtrl` derives from `wxGenericAnimationCtrl` there. However with the workaround below it won't matter in those builds either.

Anyway, the way I work around things like this is to tell the generator that both classes derive from `wxControl` and then add in definitions for the methods that are missing by copying the `MethodDef` instances from the class that has them to the one that doesn't. If `wxAnimationCtrlBase` was public (in the interface headers) then it wouldn't be necessary, but I agree that it doesn't really need to be in the documentation so the workaround is fine.

> You create it by using this control itself

That's okay. I guess I was mainly concerned about the change in usage, but if it's intended then I'm okay with it. Adding the static method would probably be a good idea in case there is some use-case we are not thinking of that needs to have the animation before the control can be created.

RobinD42 commented on src/common/animatecmn.cpp:
```diff
@@ -28,14 +28,118 @@
+
+#if !defined(wxHAS_NATIVE_ANIMATIONCTRL)
+    // In this case the "native" ctrl is the generic ctrl. See wx/animate.h
+    wxIMPLEMENT_CLASS(wxAnimationCtrl, wxControl);
```
```suggestion
    wxIMPLEMENT_CLASS(wxAnimationCtrl, wxGenericAnimationCtrl);
```

This is needed in generic-only builds when the program uses `wxAnimationCtrl` so the `IsCompatibleWith` works correctly.

vadz replied:
Will fix, thanks!

vadz committed:
Fix wxRTTI for wxAnimationCtrl when using generic implementation

It must use wxGenericAnimationCtrl as the base class in wxRTTI, as
otherwise wxAnimation::IsCompatibleWith() would fail in this case for
the generic implementation of wxAnimation.

vadz commented:
Thanks for the explanations!

The only other solution I see is to inherit `wxGenericAnimationCtrl` from `wxControl` there and duplicate all the methods. This would be a more fair representation of the actual classes and should work for wxPython too, but at the cost of duplicating ~100 lines of the docs... Not sure if it's worth it, but if you feel it is, please feel free to do it (once I push my changes soon).

vadz committed:
Remove redundant checks for IsOk() in wxAnimationGenericImpl

We already check that IsOk() returns true before calling these methods,
there is no need to do it again inside them.

Generally speaking, private functions may rely on public ones doing the
precondition checking.

vadz committed:
Add static wxAnimationCtrl::CreateCompatibleAnimation()

This may be more convenient to use than CreateAnimation() if there is no
wxAnimationCtrl object at hand.

vadz commented:
I've made the changes mentioned above and don't plan to do anything else here, so please let me know if this works for you so that we could merge it. TIA!

RobinD42 commented:
Looks good to me. Merge when ready.

vadz commented:
Done now, thanks again!
